### ISSUE TITLE
Inclusion of FrameDataProcessor in boxcars?

### ISSUE THREAD

**Issue by colonelpanic8** [nickbabcock/boxcars#150]
Hi @nickbabcock, sorry for filing an issue about this -- I do realize that its not exactly the most appropriate medium in which to have this conversation but I wasn't sure how else to contact you.

As I'm sure you're at least somewhat aware, the representation of game state provided by the rocket league replay format is a bit strange. I suspect that the replay representation corresponds very directly to the way that things are represented in unreal engine in the actual game, but for many use cases, this representation is really cumbersome and difficult to work with.

This is (part of) the reason for the existence of carball (https://github.com/SaltieRL/carball) library, as it basically takes the representation spit out by boxcars and strips away some of the nastier details (especially the way things is represented in different actors). I spent some time resuscitating carball (its in a pretty sorry state and it has some protobuf generation stuff that doesn't play well with modern python tooling), and carving out only the parts that are really necessary to transform the replay into something useful, and i did get it working (you can find that https://github.com/CUBTeamRocket/carball-lite).

However, after noticing that it chokes/loses data in certain cases, I started reading its code to try to fix these issues and realized that it is a needlessly complicated tangled mess that represents things in very inefficient ways internally (so as to make it horribly slow), violates the single responsibility principle flagrantly, and introduces completely unnecessary abstractions.

This compelled me to decide to start from scratch and write something that does what carball does (specifically just the conversion of the replay into a format where its easy to look up e.g. each players state at any given time). I've made quite a bit of progress on this (see https://github.com/IvanMalison/boxcars/commit/1fba7a4fa9fad6d8080000acbc770e2578b338c3), but a few things still need work.

tl;dr - I'm wondering if this is the kind of thing that you might want included in boxcars itself, or if I should simply make a separate crate that depends on boxcars for this. I can see arguments for both courses of action, and so I thought I'd ask you to see what you think. I obviously understand if you'd rather not include something like this in boxcars, but I do think that almost anyone dealing with the frames data is going to need to have to do something like this to the date, so I thought it might also be nice to have it built in.

Also -- since my guess is that you probably have as good of an understanding of the replay format as any non-psyonix employee out there, I'd totally appreciate thoughts/input you have on what I'm working on, and in particular any of your understanding of how boost is represented. From what I can work out from how carball handles it, you basically have to build up a derived representation where you estimate the boost total by subtracting from the value found in the replay whenever the boost is considered active. This feels super gross and I'm really hoping that the carball folks were just wrong about this, but it seems like there might simply be no other way.

**Comment by nickbabcock** [comment]
Welcome! Glad to see such an interest

Yeah I'm fine if a data frame processor is included, though I'd put it behind an opt-in feature, so that the parser and processor have a bit of separation. 

> Also -- since my guess is that you probably have as good of an understanding of the replay format as any non-psyonix employee out there, I'd totally appreciate thoughts/input you have on what I'm working on, and in particular any of your understanding of how boost is represented. From what I can work out from how carball handles it, you basically have to build up a derived representation where you estimate the boost total by subtracting from the value found in the replay whenever the boost is considered active. This feels super gross and I'm really hoping that the carball folks were just wrong about this, but it seems like there might simply be no other way.

Unfortunately, my expertise is not in interpretation. I'd refer out to either ballchasing's or bakkes's community

---

For a bit of trivia: carball predates boxcars, and iirc part of boxcars structure is tailored to making it easier to consume within carball. For instance, in a [`Frame`](https://docs.rs/boxcars/0.9.8/boxcars/struct.Frame.html) new, updated, and removed actors are separated, which loses the order of how they appear in the frame. I don't think this decision has proved impactful, but gives an idea of how there is some "lossy-ness" in boxcars parsing.

**Comment by IvanMalison** [comment]
I elected to make a separate package for now because I'm iterating really quickly and I don't want to have to bug you about release etc. I think if it stablizes i might reconsider, but for now that seems like the better approach.

Lives here, for reference: https://github.com/CUBTeamRocket/boxcar-frames

**Comment by nickbabcock** [comment]
Makes sense, and I can do a patch release with your recent additions to make it more convenient for you

**Comment by IvanMalison** [comment]
> Makes sense, and I can do a patch release with your recent additions to make it more convenient for you

that would be awesome.

**Comment by nickbabcock** [comment]
Released!

**Comment by nickbabcock** [comment]
I'm going to close this, defer to data frame support outside of boxcars.

**Comment by colonelpanic8** [comment]
yep, meant to do so myself. I moved what I made to https://github.com/rlrml/subtr-actor . I think it makes more sense there as its own thing, but it might be worth mentioning it in the README or something like that since this question does come up from time to time.

**Comment by nickbabcock** [comment]
Good thought. I've updated the readme.
